cmd/quota.go

```
@@ -0,0 +1,254 @@
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/minio/minio/cmd/config"
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/env"
+	"github.com/minio/minio/pkg/event"
+	"github.com/minio/minio/pkg/madmin"
+)
+
+// BucketQuotaSys - map of bucket and quota configuration.
+type BucketQuotaSys struct {
+	sync.RWMutex
+	quotaMap map[string]madmin.BucketQuota
+}
+
+// Set - set quota configuration.
+func (sys *BucketQuotaSys) Set(bucketName string, q madmin.BucketQuota) {
+	sys.Lock()
+	sys.quotaMap[bucketName] = q
+	sys.Unlock()
+}
+
+// Get - Get quota configuration.
+func (sys *BucketQuotaSys) Get(bucketName string) (q madmin.BucketQuota, ok bool) {
+	sys.RLock()
+	defer sys.RUnlock()
+	q, ok = sys.quotaMap[bucketName]
+	return
+}
+
+// Remove - removes quota configuration.
+func (sys *BucketQuotaSys) Remove(bucketName string) {
+	sys.Lock()
+	delete(sys.quotaMap, bucketName)
+	sys.Unlock()
+}
+
+// Exists - bucketName has Quota config set
+func (sys *BucketQuotaSys) Exists(bucketName string) bool {
+	sys.RLock()
+	_, ok := sys.quotaMap[bucketName]
+	sys.RUnlock()
+	return ok
+}
+
+// Keys - list of buckets with quota configuration
+func (sys *BucketQuotaSys) Keys() []string {
+	sys.RLock()
+	defer sys.RUnlock()
+	var keys []string
+	for k := range sys.quotaMap {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// NewBucketQuotaSys returns initialized BucketQuotaSys
+func NewBucketQuotaSys() *BucketQuotaSys {
+	return &BucketQuotaSys{
+		quotaMap: map[string]madmin.BucketQuota{},
+	}
+}
+
+// parseBucketQuota parses BucketQuota from json
+func parseBucketQuota(data []byte) (quotaCfg madmin.BucketQuota, err error) {
+	err = json.Unmarshal(data, &quotaCfg)
+	if err != nil {
+		return
+	}
+	if !quotaCfg.Type.IsValid() {
+		return quotaCfg, fmt.Errorf("Invalid quota type %s", quotaCfg.Type)
+	}
+	return
+}
+
+func enforceBucketQuota(ctx context.Context, bucket string, size int64) error {
```

Is negative value of size is allowed here ? (I think Object Layer never receives -1 as size but let's return an error when size < 0.

---

just returning here if size < 0, handlers return a missing content length header error so that should be covered.